nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/DeduplicateRecords.java

```diff
@@ -0,0 +1,690 @@
+            return useHashSet
+                ? new HashSetFilterWrapper(new HashSet<>(filterCapacity))
+                : new BloomFilterWrapper(BloomFilter.create(
+                    Funnels.stringFunnel(Charset.defaultCharset()),
+                    filterCapacity,
+                    context.getProperty(BLOOM_FILTER_FPP).asDouble()
+                ));
+        } else {
+            return new DistributedMapCacheClientWrapper(mapCacheClient);
+        }
+    }
+
+    public static final char JOIN_CHAR = '~';
+
+    @Override
+    public void onTrigger(final ProcessContext context, final ProcessSession session) throws ProcessException {
+        final FlowFile flowFile = session.get();
+        if (flowFile == null) {
+            return;
+        }
+
+        final ComponentLog logger = getLogger();
+
+        FlowFile nonDuplicatesFlowFile = session.create(flowFile);
+        FlowFile duplicatesFlowFile = session.create(flowFile);
+
+        try {
+            final long now = System.currentTimeMillis();
+
+            final FilterWrapper filter = getFilter(context);
+
+            final String recordHashingAlgorithm = context.getProperty(RECORD_HASHING_ALGORITHM).getValue();
+            final MessageDigest messageDigest = recordHashingAlgorithm.equals(NONE_ALGORITHM_VALUE.getValue())
+                    ? null
+                    : DigestUtils.getDigest(recordHashingAlgorithm);
+            final Boolean matchWholeRecord = context.getProperties().keySet().stream().noneMatch(p -> p.isDynamic());
+
+            final InputStream inputStream = session.read(flowFile);
+
+            final RecordReader reader = readerFactory.createRecordReader(flowFile, inputStream, logger);
+
+            final RecordSchema writeSchema = writerFactory.getSchema(flowFile.getAttributes(), reader.getSchema());
+
+            final OutputStream nonDupeStream = session.write(nonDuplicatesFlowFile);
+            final OutputStream dupeStream = session.write(duplicatesFlowFile);
+
+            final RecordSetWriter nonDuplicatesWriter = writerFactory.createWriter(getLogger(), writeSchema, nonDupeStream, nonDuplicatesFlowFile);
+            final RecordSetWriter duplicatesWriter = writerFactory.createWriter(getLogger(), writeSchema, dupeStream, duplicatesFlowFile);
+
+            nonDuplicatesWriter.beginRecordSet();
+            duplicatesWriter.beginRecordSet();
+            Record record;
+
+            long index = 0;
+            while ((record = reader.nextRecord()) != null) {
+                String recordValue;
+
+                if (matchWholeRecord) {
+                    recordValue = Joiner.on(JOIN_CHAR).join(record.getValues());
+                } else {
+                    recordValue = executeDynamicRecordPaths(context, record, flowFile);
+                }
+
+                String recordHash = messageDigest != null
+                        ? Hex.encodeHexString(messageDigest.digest(getBytesUtf8(recordValue)))
```

Comment:
According to the [MessageDigest](https://docs.oracle.com/javase/8/docs/api/java/security/MessageDigest.html#digest-byte:A-) documentation, calling `digest(byte[] bytes)` updates the internal hash. This means that the digest value is cumulative, as opposed to just computing and returning the value based on the current bytes.

To avoid having to create a new instance of `MessageDigest` for each record, the `digest(byte[] bytes, int offset, int length)` method can be used, as it resets the internal state so that the instance can be used again.

Reply:
`reset()` does something equivalent so I went with that.